packages/cli/src/commands/hydrogen/env/push__unstable.ts

```diff
@@ -0,0 +1,254 @@
+  const compareableRemoteVars =
+    remoteVars
+      .sort((a, b) => a.key.localeCompare(b.key))
+      .map(({key, value}) => `${key}=${value}`)
+      .join('\n') + '\n';
+
+  const compareableLocalVars =
+    Object.keys(localVariables)
+      .sort((a, b) => a.localeCompare(b))
+      .reduce((acc, key) => {
+        const {isSecret, readOnly} =
+          environmentVariables.find((variable) => variable.key === key) ?? {};
+        if (isSecret || readOnly) return acc;
+        return [...acc, `${key}=${localVariables[key]}`];
+      }, [] as string[])
+      .join('\n') + '\n';
+
+  // Confirm changes and show a generate diff of changes
+  if (!validatedEnvironment.name)
+    throw new AbortError('Missing environment name');
+
+  // Generate a list of remote read only or secrets
+  const remoteReadOnlyOrSecrets = environmentVariables.reduce(
+    (acc, {isSecret, readOnly, key}) => {
+      if (!isSecret && !readOnly) return acc;
+      const localVar = localVariables[key];
+      const remoteVar = environmentVariables.find(
+        (variable) => variable.key === key,
+      );
+      if (localVar === remoteVar?.value) return acc;
+      return [...acc, key];
+    },
+    [] as string[],
+  );
+
+  if (compareableLocalVars === compareableRemoteVars) {
+    renderInfo({
+      body: `No changes to your environment variables.${
+        Boolean(remoteReadOnlyOrSecrets.length)
+          ? `\n\nVariables that are read only or contain secret values cannot be pushed from the CLI: ${remoteReadOnlyOrSecrets.join(
+              ', ',
+            )}.`
+          : ''
+      }`,
+    });
+    return;
+  } else {
+    const diff = diffLines(compareableRemoteVars, compareableLocalVars);
+    const confirmPush = await renderConfirmationPrompt({
+      confirmationMessage: `Yes, confirm changes`,
+      cancellationMessage: `No, make changes later`,
+      message:
+        outputContent`We'll make the following changes to your environment variables for ${
+          validatedEnvironment.name
+        }:
+
+${outputToken.linesDiff(diff)}${
+          Boolean(remoteReadOnlyOrSecrets.length)
+            ? `\nRead only and secret keys cannot be pushed: ${remoteReadOnlyOrSecrets.join(
+                ', ',
+              )}\n`
+            : ''
+        }
+Continue?`.value,
+    });
+
+    // Cancelled making changes
+    if (!confirmPush) return;
+  }
+
+  if (!validatedEnvironment.id) throw new AbortError('Missing environment ID');
+  const {userErrors} = await pushStorefrontEnvVariables(
+    session,
+    config.storefront.id,
+    validatedEnvironment.id,
+    Object.entries(localVariables).map(([key, value]) => ({key, value})),
+  );
+
+  if (userErrors.length) {
+    renderWarning({
+      headline: 'Failed to upload and save environment variables',
+      body: userErrors[0]?.message,
+    });
+  }
+
+  renderSuccess({
+    body: `Environment variables push to ${
+      validatedEnvironment.branch ?? 'Preview'
+    } was successful.`,
+  });
```

Comment:
I think I just missed using AbortError on this one too! I'll adjust the wording for now until Graham gets into it. I don't believe this can be successful and have user errors.